### ISSUE TITLE
Precision-recall curve and P@N

### ISSUE THREAD

**Issue by CatarinaPC** [thunlp/OpenNRE#278]
Looking at the code on the `eval()` function, `data_loader.py` script in the ` BagREDataset` class:

```
def eval(self, pred_result):
        """
        Args:
            pred_result: a list with dict {'entpair': (head_id, tail_id), 'relation': rel, 'score': score}.
                Note that relation of NA should be excluded.
        Return:
            {'prec': narray[...], 'rec': narray[...], 'mean_prec': xx, 'f1': xx, 'auc': xx}
                prec (precision) and rec (recall) are in micro style.
                prec (precision) and rec (recall) are sorted in the decreasing order of the score.
                f1 is the max f1 score of those precison-recall points
        """
        sorted_pred_result = sorted(pred_result, key=lambda x: x['score'], reverse=True)
        prec = []
        rec = []
        correct = 0
        total = len(self.facts)
        for i, item in enumerate(sorted_pred_result):
            if (item['entpair'][0], item['entpair'][1], item['relation']) in self.facts:
                correct += 1
            prec.append(float(correct) / float(i + 1))
            rec.append(float(correct) / float(total))
        auc = sklearn.metrics.auc(x=rec, y=prec)
        np_prec = np.array(prec)
        np_rec = np.array(rec) 
        f1 = (2 * np_prec * np_rec / (np_prec + np_rec + 1e-20)).max()
        mean_prec = np_prec.mean()
        return {'micro_p': np_prec, 'micro_r': np_rec, 'micro_p_mean': mean_prec, 'micro_f1': f1, 'auc': auc}

```

If I want to show the **precision-recall curve** do I just need plot `np_prec` on the y axis and `np_rec` on the x axis (without the need of another function) ?

The same goes for computing **P@N** (Precision@N), for different N's, e.g., N=100, N=200, N=300. Do I just need to look at the index N in the `np_prec` array?

**Comment by gaotianyu1350** [comment]
1. Right. You should plot (np_rec, np_prec).

2. Right. np_prec[99] means P@100.

**Comment by CatarinaPC** [comment]
I was hoping to get some feedback.

I trained a model on data which I obtained by distant supervision (DBpedia + Wikipedia). I divided it in train, test, dev and train a bag-level PCNN+ATT model using OpenNRE.

I'm getting super good results, which is suspicious:

2020-08-30 03:35:38,029 - root - INFO - Test set results:
2020-08-30 03:35:38,029 - root - INFO - AUC: 0.9710321405860297
2020-08-30 03:35:38,029 - root - INFO - Micro F1: 0.9178284843322724
2020-08-30 03:35:38,029 - root - INFO - P@100: 1.0
2020-08-30 03:35:38,029 - root - INFO - P@200: 1.0
2020-08-30 03:35:38,029 - root - INFO - P@300: 1.0

Evaluating on a chunk of the same data as the training can cause the results to be this good? How can I get more accurate results keeping the training the same? Should I create a small manual test set (with bags or without?) ?

**Comment by gaotianyu1350** [comment]
Maybe you can first check whether there is overlap between train and test sets. The result is indeed "too good" and I feel like there is a test set leak problem.

If nothing is wrong with the dataset split, maybe use a larger test set.

**Comment by CatarinaPC** [comment]
I ran this code:

```
import json

with open("../../datasets/processed/v1/train_sentences.txt", 'r') as handle:
    train_data = [json.loads(line) for line in handle]

with open("../../datasets/processed/v1/test_sentences.txt", 'r') as handle:
    test_data = [json.loads(line) for line in handle]

train_text = set([td['text'] for td in train_data])
test_text = set([td['text'] for td in test_data])

intersec = train_text.intersection(test_text)
print(len(intersec))

train_triple = set([td['relation']+"_"+td['h']["id"]+"_"+td['t']["id"] for td in train_data])
test_triple = set([td['relation']+"_"+td['h']["id"]+"_"+td['t']["id"] for td in test_data])

intersec = train_triple.intersection(test_triple)
print(len(intersec))

train_tuple_text = set([td['h']["id"]+"_"+td['t']["id"]+"_"+td["text"] for td in train_data])
test_tuple_text = set([td['h']["id"]+"_"+td['t']["id"]+"_"+td["text"] for td in test_data])

intersec = train_tuple_text.intersection(test_tuple_text)
print(len(intersec))
```
I first checked there are sentences repeated between the train and test sets. I got `3785`.
However this can happen for 2 reasons:

- 2 entities may have more than one relationship between them, (ent1,ent2,rel1) and (ent1, ent2, rel2), and one of them is in the train set and the other on the test set. Due to the distant supervision algorithm to build the dataset, the same example sentences wll be used for both triples but have different relationships assigned.

- a sentence may have more than one relationship in it. For example "Obama was born in Honululu, Hawaii". this sentence will be used for (Obama,Honolulu, birthPlace) and (Honululu, Hawaii, city) (something like that). And one of this triples be in the train set and the other on the test set. But again, like in the previous point, this is not a leak, this does not "help" the model, just "confuses" it

So I checked if there were repeated triples (ent1, ent2, rel) and got `0`.

Then, checked repeated (ent1, ent2) and sentences and got `618`, which can happen for the first reason I presented.

I don't think there is a leak. What do you think?

**Comment by CatarinaPC** [comment]
1. How can I compute or change the code to compute typical metrics like precision, recall and f1-score per class/ relationship type, when using a bag-level model?

I see in the code that in bag level evaluation there is a `pred_result` list to which the predictions (one for each relation type) of each entity pair are added, including the ones that do not have the highest confidence (which we typically refer to just as prediction, and get as a result when using infer() ).

2. Is this because it is necessary when computing AUC?
3. And, with these added predictions for every relation type is it possible to compute the metrics I asked in the first question?

**Comment by gaotianyu1350** [comment]
The evaluation code is in `opennre/framework/data_loader.py:230`.

To calculate the precision-recall curve and the AUC, all prediction results (not only those with the highest scores) are needed. Each point at the precision-recall curve actually represents one of these results. Usually we cannot compute precision or recall for this setting, because there are so many (each point is a precision-recall pair). But we can report an f1 score which usually refers to the highest f1 score of all points.

On the other hand, you can report precision@N indicating the precision for the top-N prediction results.

**Comment by CatarinaPC** [comment]
Thank you. I get precision@100, precision@200 and precision@300 = 1
The results are all extremely high. I have AUC = 0.9710 and micro-f1 of 0.9178.

So to report by class/relationship type it is not possible. Not even if we just use the prediction results with the highest scores, somehow?

**Comment by gaotianyu1350** [comment]
Report by class/relation is still possible. The way of calculation is the same, just filter out predictions of other relations.

**Comment by CatarinaPC** [comment]
https://github.com/thunlp/OpenNRE/blob/60a8ceb42e1cfacbde3c8cfb5f758fb7fe96bdc4/opennre/framework/data_loader.py#L230

So you mean, for each relation remove from `sorted_pred_result` [here](https://github.com/thunlp/OpenNRE/blob/60a8ceb42e1cfacbde3c8cfb5f758fb7fe96bdc4/opennre/framework/data_loader.py#L241) all other relations and do the same computations?

**Comment by gaotianyu1350** [comment]
Right.
